libia2/ia2.c

```diff
@@ -21,65 +41,81 @@ int protect_pages(struct dl_phdr_info *info, size_t size, void *data) {
     Elf64_Addr start = info->dlpi_addr + phdr.p_vaddr;
     Elf64_Addr end = start + phdr.p_memsz;
     if (start <= address && address <= end) {
-      this_compartment = true;
-      break;
+      return true;
     }
   }
-  if (!this_compartment) {
+  return false;
+}
+
+/// Map ELF segment flags to mprotect access flags
+static int segment_flags_to_access_flags(Elf64_Word flags) {
+  return ((flags & PF_X) != 0 ? PROT_EXEC : 0) |
+         ((flags & PF_W) != 0 ? PROT_WRITE : 0) |
+         ((flags & PF_R) != 0 ? PROT_READ : 0);
+}
+
+int protect_pages(struct dl_phdr_info *info, size_t size, void *data) {
+  if (!data || !info) {
+    printf("Passed invalid args to dl_iterate_phdr callback\n");
+    exit(-1);
+  }
+
+  struct PhdrSearchArgs *search_args = (struct PhdrSearchArgs *)data;
+  Elf64_Addr address = (Elf64_Addr)search_args->address;
+  if (!in_loaded_segment(info, address)) {
+    // Continue iterating to check the next object
     return 0;
   }
-  const char *ignored_sections[NUM_IGNORED_SECTIONS][2] = {
-      {"__start_ia2_shared_data", "__stop_ia2_shared_data"},
-      {"__start_ia2_shared_rodata", "__stop_ia2_shared_rodata"},
-  };
-  struct AddressRange {
-    uint64_t start;
-    uint64_t end;
-  };
-  struct AddressRange ignored_ranges[NUM_IGNORED_SECTIONS];
+
   void *lib = dlopen(info->dlpi_name, RTLD_NOW);
-  if (!lib) {
+  if (!lib)
     exit(-1);
-  }
-  for (size_t i = 0; i < NUM_IGNORED_SECTIONS; i++) {
+
+  struct AddressRange shared_ranges[NUM_SHARED_RANGES] = {0};
+  struct AddressRange *cur_range = &shared_ranges[0];
+  for (size_t i = 0; i < NUM_SHARED_RANGES; i++) {
     // Clear any potential old error conditions
     dlerror();
-    ignored_ranges[i].start = (uint64_t)dlsym(lib, ignored_sections[i][0]);
-    uint64_t aligned_start = ignored_ranges[i].start & ~0xFFFUL;
-    if (aligned_start != ignored_ranges[i].start) {
-      printf("Start of section %s is not page-aligned\n",
-             ignored_sections[i][0]);
-      exit(-1);
+
+    cur_range->start = (uint64_t)dlsym(lib, shared_sections[i][0]);
+    if (!cur_range->start) {
```

Comment:
We should check `dlerror` rather than comparing with `NULL`. Though I'm not sure how relevant getting `NULL` from `dlsym` is for start symbols of sections in `PT_LOAD` segments.

Reply:
While I agree in general, we don't want to accept NULL values for either of these symbols, right?

Reply:
I think you're right that we don't want to accept `NULL` values, but I remember @endbr64 ran into issues with this. I think we were modifying the ld script and it made one of the symbols `NULL` so splitting this into two separate checks would be good to easily distinguish between the symbol not existing and the symbol having an invalid value.

Reply:
fwiw I'd like to avoid modify ld scripts in weird ways to avoid the root issue in that case, but splitting this into two checks would make debugging easier.

Reply:
Done. I left the first one as a null check because we want to silently continue in either case, though.